coderd/provisionerjobs.go

```diff
@@ -334,98 +221,223 @@ func convertProvisionerJob(provisionerJob database.ProvisionerJob) codersdk.Prov
+func newLogFollower(
+	ctx context.Context, logger slog.Logger, db database.Store, pubsub database.Pubsub,
+	job database.ProvisionerJob, after int64, r *http.Request, rw http.ResponseWriter,
+) *logFollower {
+	return &logFollower{
+		ctx:      ctx,
+		logger:   logger,
+		db:       db,
+		pubsub:   pubsub,
+		r:        r,
+		rw:       rw,
+		jobID:    job.ID,
+		after:    after,
+		complete: jobIsComplete(logger, job),
+		kicks:    make(chan provisionersdk.ProvisionerJobLogsNotifyMessage),
+		errors:   make(chan error),
+	}
+}
 
-	sendLog := func(log *database.ProvisionerJobLog) {
-		select {
-		case bufferedLogs <- log:
-			logger.Debug(context.Background(), "subscribe buffered log", slog.F("stage", log.Stage))
-			lastSentLogID.Store(log.ID)
-		default:
-			// If this overflows users could miss logs streaming. This can happen
-			// we get a lot of logs and consumer isn't keeping up.  We don't want to block the pubsub,
-			// so just drop them.
-			logger.Warn(context.Background(), "provisioner job log overflowing channel")
+func (f *logFollower) follow() {
+	// note that we only need to subscribe to updates if the job is not yet
+	// complete.
+	if !f.complete {
+		subCancel, err := f.pubsub.SubscribeWithErr(
+			provisionersdk.ProvisionerJobLogsNotifyChannel(f.jobID),
+			f.listener,
+		)
+		if err != nil {
+			httpapi.Write(f.ctx, f.rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "failed to subscribe to job updates",
+				Detail:  err.Error(),
+			})
+			return
 		}
+		defer subCancel()
+
+		// we were provided `complete` prior to starting this subscription, so
+		// we also need to check whether the job is now complete, in case the
+		// job completed between the last time we queried the job and the start
+		// of the subscription.  If the job completes after this, we will get
+		// a notification on the subscription.
+		job, err := f.db.GetProvisionerJobByID(f.ctx, f.jobID)
+		if err != nil {
+			httpapi.Write(f.ctx, f.rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "failed to query job",
+				Detail:  err.Error(),
+			})
+			return
+		}
+		f.complete = jobIsComplete(f.logger, job)
 	}
 
-	closeSubscribe, err := api.Pubsub.Subscribe(
-		provisionerJobLogsChannel(jobID),
-		func(ctx context.Context, message []byte) {
-			if endOfLogs.Load() {
-				return
+	var err error
+	f.conn, err = websocket.Accept(f.rw, f.r, nil)
+	if err != nil {
+		httpapi.Write(f.ctx, f.rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Failed to accept websocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	defer f.conn.Close(websocket.StatusNormalClosure, "done")
+	go httpapi.Heartbeat(f.ctx, f.conn)
+
+	// query for logs once right away, so we can get historical data from before
+	// subscription
+	if err := f.query(); err != nil {
+		if f.ctx.Err() == nil && !xerrors.Is(err, io.EOF) {
+			// neither context expiry, nor EOF, close and log
+			f.logger.Error(f.ctx, "failed to query logs", slog.Error(err))
```

Comment:
Should have job ID as field

Reply:
HTTP handler sets `job_id` in the logger before passing it to the logFollower